### PR TITLE
Add labels support to terraform

### PR DESCRIPTION
**Pull request by ptak7**
## PR Description
This PR adds support to labels on 
- Services
- SLOs
- (Project Code for provider is also added but for now it's not supported by n9)

## Labels structure
```
...
  label {
    key = "team"
    values = ["green","sapphire"]
  }

  label {
    key = "env"
    values = ["dev", "staging", "prod"]
  }
...
```
A decision to use such structure was made because there's no support for a nested map of lists in terraform SDK.
This is also similar to how Alert Conditions are defined so we are consequent with the selected approach.

## How to test locally
Add ```replace github.com/nobl9/nobl9-go v0.2.1 => ../nobl9-go``` to `go.mod`
Run ```make install``` in root directory
Initialize terraform and run ```terraform apply``` with yaml file

## How to test remotely
Fill `okta_auth_server`, `client_id` and `client_secret` with proper credentials.
Apply specific `main.tf` file via `terraform apply`. Example file:
```
terraform {
  required_providers {
    nobl9 = {
      source = "nobl9.com/nobl9/nobl9"
      version = "0.4.2"
    }
  }
}

provider "nobl9" {
  organization      = "nobl9-dev"
  project           = "default"
  ingest_url        = "https://dev-sapphire.nobl9.dev/api"
  okta_org_url      = "https://accounts.nobl9.dev"
  okta_auth_server  = "***"
  client_id         = "***"
  client_secret     = "***"
}

locals {
  latency_objective_sets = {
    "Ok" : {
      value = 4000
    },
    "Good" : {
      value = 3000
    },
    "Well" : {
      value = 2000
    },
  }
}

resource "nobl9_project" "test-labels-project" {
  name = "test-labels-project"
  labels = ["resource:project", "team:grey", "team:green", "env:dev", "team:purple", "kto:ja", "kto:ty", "team:sapphire"]
}

resource "nobl9_service" "test-labels-service" {
  name = "test-labels-service"
  labels = ["resource:service", "team:grey", "team:green", "env:dev", "team:purple", "kto:ja", "kto:ty", "team:sapphire"]
  project         = "test-labels-project"
}

resource "nobl9_slo" "test-labels-slo" {
  name             = "test-labels-slo"
  service          = "test-labels-service"
  budgeting_method = "Occurrences"
  project         = "test-labels-project"
  labels = ["team:grey", "team:green", "env:dev", "team:purple", "kto:ja", "kto:ty", "team:sapphire"]
  time_window {
    unit       = "Day"
    count      = 14
    is_rolling = true
  }
  dynamic "objective" {
    for_each = local.latency_objective_sets
    content {
      target       = 0.999
      display_name = objective.key
      value        = objective.value["value"]
      op           = "lte"
      raw_metric {
        query {
          cloudwatch {
            dimensions {
              name  = "ApiName"
              value = "value.api.name"
            }
            metric_name = "Latency"
            namespace   = "AWS/ApiGateway"
            region      = "eu-central-1"
            stat        = "p99"
          }
        }
      }
    }
  }

  indicator {
    name    = "cloudwatch-direct"
    kind    = "Direct"
    project = "cloudwatch-direct"
  }
}

```

### CODE REVIEW

**Comment by kskitek** [comment]
@ptak7 can we talk on why you chose to use `[]string` for labels?

- Our yamls use `map[string][]string`.
- Also looking at things like aws tags or kubernetes labels, it is a map 
  - https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/resource-tagging#getting-started-with-resource-tags
  - https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/labels#labels)

File: nobl9/metadata_schema.go [hunk 1]
```diff
@@ -51,15 +82,61 @@ func schemaDescription() *schema.Schema {
...
+type label struct {
+	key   string
+	value string
+}
+
+func newLabel(labelRaw string) (*label, error) {
+	labelSegments := strings.Split(labelRaw, ":")
+	if len(labelSegments) != 2 {
+		return nil, fmt.Errorf("wrong label format, expected \"key:value\", got \"%s\" ", labelRaw)
 	}
```
Comment (kskitek) [inline nobl9/metadata_schema.go:114]: If we want to stick with `key:value` format, we should use it as `ValidateFunc` in schema definition I think

File: nobl9/metadata_schema.go [hunk 2]
```diff
@@ -51,15 +82,61 @@ func schemaDescription() *schema.Schema {
...
+	return &label{
+		key:   labelSegments[0],
+		value: labelSegments[1],
+	}, nil
+}
+
+func (l label) toString() string {
```
Comment (kskitek) [inline nobl9/metadata_schema.go:121]: I think the good golang practice is to implement stringer interface in that case, wdyt?

```suggestion
func (l label) String() string {
```

File: nobl9/metadata_schema.go (hunk 1 above)
Comment (ptak7) [reply nobl9/metadata_schema.go:114]: aws tags and kubernetes labels are `map[string]string`, not `map[string][]string

Terraform SDK does not support a list of strings as a map values:
https://github.com/hashicorp/terraform-plugin-sdk/blob/main/helper/schema/schema.go#L2080

This was the simplest way I could think of to support duplicate label keys that we need to support in n9

File: nobl9/metadata_schema.go (hunk 2 above)
Comment (ptak7) [reply nobl9/metadata_schema.go:121]: true :)

File: nobl9/resource_slo_test.go [hunk 3]
```diff
@@ -49,7 +48,7 @@ func TestAcc_Nobl9SLO(t *testing.T) {
 			resource.Test(t, resource.TestCase{
 				PreCheck:          func() { testAccPreCheck(t) },
 				ProviderFactories: ProviderFactory(),
-				CheckDestroy:      CheckDestory("nobl9_slo", n9api.ObjectSLO),
+				//CheckDestroy:      CheckDestory("nobl9_slo", n9api.ObjectSLO),
```
Comment (kskitek) [inline nobl9/resource_slo_test.go:51]: why this one is commented?

**Comment by klaudia-nobl9** [comment]
During the testing I found that removing labels doesn't work properly. 

After creating resource with 1 label and trying to reapply that resource without a label, I get a message: `No changes. Your infrastructure matches the configuration.` Same thing happens with **more than 1 label** when trying to remove **all** labels from the resource.
On the other hand, after creating resource with 2 labels and trying to reapply that resource but only with 1 label this time, apply is completed successfully.

Basically the scenario where I **remove all labels from the resource** doesn't work.

**Review by mkaras-nobl9 — COMMENTED** [review]
looks good :1st_place_medal: after tests fix ready to merge

**Review by mkaras-nobl9 — APPROVED** [review]
:rocket:
